### ISSUE TITLE
[Tracking] MOSS-TTS-Local streaming vocoder: full stateful CUDA graph (blocked on codec KV-cache fix)

### ISSUE THREAD

**Issue by JiaxinD** [sgl-project/sglang-omni#801]
## Goal

Capture the full **stateful** MOSS-TTS-Local streaming vocoder decode (`_CodecStreamSession.step` to
`codec._decode_frame`) into CUDA graphs and replay it, instead of re-launching the per-frame kernel storm.
Under streaming the codec decodes one tiny chunk per scheduler tick, so it is launch-bound: the target is a
large per-step speedup. Measured ~2.87x on the per-step decode (T=5, B=8: eager 53.3 ms vs graph replay
18.5 ms) when run without the blocker below.

This issue tracks the PRs and discussion for the vocoder CUDA graph direction. Current status: **blocked on a
codec-side KV-cache storage fix** (see Blocker).

## State distribution (measured, confirmed with Gary)

Where the cross-step streaming state actually lives in the decode path. From running one `_decode_frame` step
and checking which `_streaming_state` modules mutate (98 of them, all under `decoder`):

- **Decoder attention x92** (`MossAudioTokenizerMultiheadAttention` / `MHAState`): each holds the streaming
  KV cache (`cached_keys` / `cached_values` / `cached_positions`) plus a RoPE `offset`.
- **Decoder transformer x6** (`MossAudioTokenizerTransformer` / `TransformerState`): each holds `offsets`
  (a position counter).
- **Quantizer** (`MossAudioTokenizerResidualLFQ`) is **stateless**: not a StreamingModule (no
  `_streaming_state`), `decode_codes` is a pure lookup.
- **conv / upsampling**: no streaming state.

So a full stateful vocoder graph has exactly **one** state-mechanism blocker (the KV cache), not many. The
`offset` / `offsets` (92 + 6) are already in-place tensor updates and are graph-safe.

## Blocker / prerequisite (codec team)

The streaming attention KV cache is stored by **reassigning the state attribute to a freshly allocated tensor
every step**, so its `data_ptr` changes each step. A CUDA graph records ops on fixed addresses and does not
follow a Python-level attribute rebind, so replay keeps reading the capture-time (empty) cache. This is the
one thing that blocks stateful graph capture (with state restored, graph replay diverges from eager by
maxdelta ~0.3 to 0.96 on real codes: step 0 exact, step 1 onward stale).

`_update_streaming_cache` (codec `modeling_moss_audio_tokenizer.py`, line numbers approximate, snapshot
`f6e20e5`):

```python
# ~line 883, bounded-context path
state.cached_keys = torch.where(exec_mask, new_cached_k, cached_k)
```

What the graph needs from the cache: a step-to-step **stable storage address** (a fixed pre-allocated buffer
written in place, not rebound) plus bit-identity. Decoder attention context is bounded (`{125, 250, 400}`),
so the cache shape is already fixed; an in-place `scatter_` / `index_copy_` write is likely lighter than
copying the whole buffer. The codec already contains that pattern in `RingKVCache.complete()` (fixed
pre-allocated `self.cache`, in-place `scatter_` / `index_copy_`, in-place `end_offset`), though `RingKVCache`
appears unused in the modeling file (no instantiation, no `complete()` caller).

- [ ] https://github.com/sgl-project/sglang-omni/pull/811

## CUDA graph implementation (my part, after the blocker is unblocked)

- [x] CUDA graph capture / replay framework: PR #798 (currently **on hold**).
      The current #798 implementation made the codec run stateless, which made the bit-identity gate
      tautological (both arms stateless) and degraded streaming audio at chunk boundaries. It is on hold
      until the codec cache is fixed, rather than shipped.
- [ ] After the codec cache is unblocked: restore streaming state, wire the graph back in, validate the real
      gate.
- [ ] End-to-end streaming audio bit-identity (no chunk-boundary artifacts).

## Acceptance criteria (the real gate, not a false success)

Once the codec cache fix and the graph are wired together, all of these must hold. The previous gate passed
only because both arms were stateless, which is not a real check:

- [ ] **Real bit-identity gate with streaming state restored**: graph replay == eager, bit-identical
      (`torch.equal`, maxdelta 0).
- [ ] **End-to-end streaming audio bit-identical** (no chunk-boundary artifacts). For reference, the stateless
      version measured ~+3 dB SNR vs ~+42 dB for the correct stateful decode, mean WER ~6% (0.058, vs 0
      for the stateful and offline decode).
- [ ] **Address stability**: cache `data_ptr` unchanged across steps.

## Related context / history

- PR #798 review (@Ratish1): the ExitStack contract question was the start of the thread that surfaced the
  root cause.
- _(placeholder for other related PRs / discussions)_

## Follow-up / roadmap

- [ ] _(roadmap items to be scoped with Chenyang; this issue stays focused on the vocoder CUDA graph)_

---

cc @gaoyang07 @Ccyest @zhaochenyang20

**Comment by gaoyang07** [comment]
I may not treat PR #811 as the final direction. It is a useful correctness-oriented intermediate step: it fixes the immediate state-addressing issue by keeping the existing attention KV tensors stable via `copy_`, preserves the streaming context correctly, and avoids the current dynamic flash-attention path by forcing SDPA. That makes it much safer than #798 implementation as submitted.

However, from a performance architecture perspective, #811 is not the optimal design. It keeps the quantizer outside the graph, pays an eager `decode_codes` launch path, copies dense decoder hidden states into static graph inputs, forces SDPA, and still updates the KV cache with full-window `where + copy_` traffic every step. This addresses CUDA Graph correctness, but it does not minimize memory bandwidth or fully collapse the vocoder step into one replay.

The original target of #798 is still the better end-state: graph the full `_decode_frame` path, including quantizer decode and decoder execution, under a real stateful streaming session. The issue is not the graph boundary itself. The root cause is in the codec streaming path: the attention KV cache is rebound to a newly allocated tensor at every step, so its `data_ptr` changes across chunks. CUDA Graph replay records fixed addresses and cannot follow that Python-level rebind, which makes true stateful replay stale.

This looks fixable on the codec side. The codec already contains a graph-safe pattern in `RingKVCache.complete`: a fixed preallocated buffer updated in place, with stable storage across steps. That path just does not appear to be wired into the current streaming attention path. If streaming KV updates are changed from tensor rebinding to in-place writes into a fixed buffer, and there is no semantic requirement that forces rebinding, then #798’s full-vocoder graph framework can be resumed and should have the better performance ceiling. @JiaxinD can then finish the implementation and report the real stateful benchmark, including the expected ~2.87x per-step decode speedup.

My recommendation:

- Hold #798 in its current form because its correctness gate is invalid.
- Use #811 as a reference for the required stateful correctness fixes, especially stable KV storage and real streaming-state tests.
- But do not choose #811 as the final architecture.
- Move #798 forward after replacing attention KV rebinding with a fixed-address ring-style cache and adding a real stateful bit-identity gate.

In short: #811 is the safer patch; #798 is the better final design once the KV cache is made graph-safe.
